drivers/SmartThings/zigbee-switch/src/frient/init.lua

```diff
@@ -0,0 +1,209 @@
+    { mfr = "frient A/S", model = "SPLZB-144" },
+    { mfr = "frient A/S", model = "SPLZB-147" },
+    { mfr = "frient A/S", model = "SMRZB-143" },
+    { mfr = "frient A/S", model = "SMRZB-153" },
+    { mfr = "frient A/S", model = "SMRZB-332" },
+    { mfr = "frient A/S", model = "SMRZB-342" },
+}
+
+local POWER_FAILURE_ALARM_CODE = 0x03
+
+
+local function alarm_report_handler(driver, device, zb_rx)
+    local alarm_status = zb_rx.body.zcl_body
+    log.error("Received alarm report:"..util.stringify_table(zb_rx.body.zcl_body, nil, true))
+    if ((alarm_status.cluster_identifier.value == SimpleMetering.ID) and (alarm_status.alarm_code.value == POWER_FAILURE_ALARM_CODE)) then
+        device.thread:call_with_delay(2, function(d)
+            device:emit_event_for_endpoint(zb_rx.address_header.src_endpoint.value, capabilities.powerSource.powerSource.unknown())
+        end
+        )
+    end
+end
+
+local function device_added(driver, device)
+end
+
+-- Handler for voltage measurement
+local function voltage_measurement_handler(driver, device, value, zb_rx)
+    local raw_value  = value.value
+    -- By default emit raw value
+    local multiplier = device:get_field(VOLTAGE_MEASUREMENT_MULTIPLIER_KEY) or 1
+    local divisor    = device:get_field(VOLTAGE_MEASUREMENT_DIVISOR_KEY) or 1
+
+    if divisor == 0 then
+        log.warn("Voltage scale divisor is 0; using 1 to avoid division by zero")
+        divisor = 1
+    end
+
+    raw_value  = raw_value * multiplier / divisor
+
+    local mult = 10 ^ 1 -- Round off to 1 decimal place
+    raw_value  = math.floor(raw_value * mult + 0.5) / mult
+
+    if device:supports_capability(capabilities.voltageMeasurement) then
+        device:emit_event_for_endpoint(zb_rx.address_header.src_endpoint.value, capabilities.voltageMeasurement.voltage({ value = raw_value, unit = "V" }))
+    end
+    if device:supports_capability(capabilities.powerSource) then
+        device:emit_event(capabilities.powerSource.powerSource.mains())
+    end
+end
+
+local function current_measurement_divisor_handler(driver, device, divisor, zb_rx)
+    log.info("current divisor value = ", divisor.value)
+    local raw_value = divisor.value
+    device:set_field(CURRENT_MEASUREMENT_DIVISOR_KEY, raw_value, { persist = true })
+end
+
+local function current_measurement_multiplier_handler(driver, device, multiplier, zb_rx)
+    log.info("current multipl value = ", multiplier.value)
+    local raw_value = multiplier.value
+    device:set_field(CURRENT_MEASUREMENT_MULTIPLIER_KEY, raw_value, { persist = true })
+end
+
+local function voltage_measurement_divisor_handler(driver, device, divisor, zb_rx)
+    log.info("V divisor value = ", divisor.value)
+
+    local raw_value = divisor.value
+    device:set_field(VOLTAGE_MEASUREMENT_DIVISOR_KEY, raw_value, { persist = true })
+end
+
+local function voltage_measurement_multiplier_handler(driver, device, multiplier, zb_rx)
+    log.info("Volt Multiplayer value = ")
```

Comment:
several log statements here that ought to be removed. This one doesn't actually log any values.